p2p/src/msg.rs

```diff
@@ -228,19 +251,50 @@ impl Writeable for MsgHeader {
 	}
 }
 
-impl Readable for MsgHeader {
-	fn read(reader: &mut dyn Reader) -> Result<MsgHeader, ser::Error> {
+impl Readable for MsgHeaderWrapper {
+	fn read(reader: &mut dyn Reader) -> Result<MsgHeaderWrapper, ser::Error> {
 		let m = magic();
 		reader.expect_u8(m[0])?;
 		reader.expect_u8(m[1])?;
-		let (t, len) = ser_multiread!(reader, read_u8, read_u64);
+
+		// Read the msg header.
+		// We do not yet know if the msg type is one we support locally.
+		let (t, msg_len) = ser_multiread!(reader, read_u8, read_u64);
+
+		// Attempt to convert the msg type byte into one of our known msg type enum variants.
+		// Check the msg_len while we are at it.
 		match Type::from_u8(t) {
-			Some(ty) => Ok(MsgHeader {
-				magic: m,
-				msg_type: ty,
-				msg_len: len,
-			}),
-			None => Err(ser::Error::CorruptedData),
+			Some(msg_type) => {
+				// TODO 4x the limits for now to leave ourselves space to change things.
+				let max_len = max_msg_size(msg_type) * 4;
+				if msg_len > max_len {
+					error!(
+						"Too large read {:?}, max_len: {}, msg_len: {}.",
+						msg_type, max_len, msg_len
+					);
+					return Err(ser::Error::TooLargeReadErr);
+				}
+
+				Ok(MsgHeaderWrapper::Known(MsgHeader {
+					magic: m,
+					msg_type,
+					msg_len,
+				}))
+			}
+			None => {
+				// Unknown msg type, but we still want to limit how big the msg is.
+				// Default to max_block_size (4x as above for space to change things).
+				let max_len = max_block_size() * 4;
+				if msg_len > max_len {
```

Comment:
Mostly unrelated, because we do the same above, but does it mean if we get a too large message we (this peer thread) will stuck forever checking the size?

Reply:
I need to check but I'm guessing what we _should_ be doing in this scenario is banning the peer.
We should _never_ receive a msg that exceeds the size limits from an honest peer.

I'm not sure we are doing this today though, but I had assumed we were.

Reply:
I assume that what happens here is - 

* we attempt to deserialize the msg header
  * 11 bytes (fixed size) read from stream
  * deserialization error due to `msg_len` exceeding `max_len`
* `try_break!` loops again
* we then attempt to deserialize the next msg header
  * 11 bytes (fixed size) read from stream
  * the magic bytes will not match as these 11 bytes are not actually a msg header
* `try_break!` is now effectively looping looking for a valid 11 byte chunk starting with magic bytes

I was actually thinking about something similar to this yesterday related to connection dropping unexpectedly. I was wondering if the sender and the receiver could get "out of alignment" as we identify msg based on the first 2 bytes (the magic number).

It looks like this could actually happen here as well.
If the next bytes are not the magic bytes then we are going to read 11 bytes at a time and if we are misaligned by even one byte we will not identify the next msg header correctly.

A couple of alternative approaches - 

1. If we don't read a full msg we drop the connection. We cannot rely on the next bytes read being the next pair of magic bytes.
2. Alternatively - consider adding some logic to read (and discard) bytes until we encounter the next pair of magic bytes? I'm not sure if this is an approach often used in practice?

Reply:
I'm going to open a separate issue to track this.

Reply:
Issue #2793.